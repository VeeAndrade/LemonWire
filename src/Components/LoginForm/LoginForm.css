form {
  display: flex;
  flex-direction: column;
  padding-top: 405px;
  padding-left: 170px;
  position: absolute;
  z-index: 3;
}

input {
  background: 0;
  border: 0;
  font-size: 2em;
  margin-bottom: 50px;
  text-indent: .5em;
  width: 350px;
}

.login-btn {
  align-self: center;
  font-size: 1em;
  height: 30px;
  margin-top: 20px;
  width: 120px;
}
