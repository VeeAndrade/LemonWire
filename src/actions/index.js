export const getMovies = movies => ({
  type: 'GET_MOVIES',
  movies
});

export const updateUser = user => ({
  type: 'UPDATE_USER',
  user
});

export const updateSelectedMovie = movie => ({
  type: 'UPDATE_SELECTED_MOVIE',
  movie
});

export const logoutUser = user => ({
  type: 'LOGOUT_USER',
  user
})

export const getRatings = ratings => ({
  type: 'GET_RATINGS',
  ratings
})

export const removeRatings = ratings => ({
  type: 'REMOVE_RATINGS',
  ratings
})
