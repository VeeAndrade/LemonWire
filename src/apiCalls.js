export const getUserInfo = user => {
  const options = {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
    .then(response => {
    if (!response.ok) {
      throw Error('200 status code not found: getUserInfo throw error')
    }
    return response.json()
  })
}

export const getUserRatings = userId => {
  return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userId}/ratings`)
  .then(response => {
    if (!response.ok) {
      throw Error('200 status code not found: getUserRatings throw error')
    }
    return response.json()
  })
}