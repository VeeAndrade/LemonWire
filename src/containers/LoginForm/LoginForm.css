form {
  align-items: center;
  display: flex;
  flex-direction: column;
  grid-area: form;
  justify-content: flex-start;
  margin: 50px;
}

input {
  border: 1px solid #EBEBEB;
  border-radius: 4px;
  box-shadow: 0 2px 4px rgba(0,0,0,0.1);
  font-size: 1.3em;
  height: 50px;
  margin-bottom: 30px;
  text-indent: .5em;
  width: 300px;
}

.error-msg {
  color: #E6D531;
  width: 10em;
}

.login-btn {
  border-radius: 4px;
  font-size: .8em;
  font-weight: bold;
  height: 2.5em;
  margin-bottom: 30px;
  width: 10em;
}

.login-btn:hover {
  background-color: #E6D531;
  cursor: pointer;
}
