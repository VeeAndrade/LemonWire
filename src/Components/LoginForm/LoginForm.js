import React, { Component } from 'react';
import './LoginForm.css';

class LoginForm extends Component {
  constructor() {
    super();
    this.state= {username: '', password: ''}
  }

  handleChange = event => {
    this.setState({[event.target.name]: event.target.value});
  }

  render() {
    return (
      <form>
        <input
          type="text"
          placeholder="Username..."
          name="username"
          value={this.state.username}
          onChange={this.handleChange}
        />
        <input
          type="text"
          placeholder="Password..."
          name="password"
          value={this.state.password}
          onChange={this.handleChange}
        />
        <button className='login-btn'>Login</button>
      </form>
    )
  }
}

export default LoginForm
