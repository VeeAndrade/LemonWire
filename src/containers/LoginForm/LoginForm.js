import React, { Component } from 'react';
import './LoginForm.css';
import { connect } from 'react-redux';
import { updateUser } from '../../actions';

class LoginForm extends Component {
  constructor() {
    super();
    this.state= {email: '', password: '', error: false}
  }

  handleChange = event => {
    this.setState({[event.target.name]: event.target.value});
  }

  verifyInputs = event => {
    event.preventDefault();
    let values = Object.values(this.state);
    values.includes('') ? this.setState({error: true}) : this.completeLogin();
  }

  completeLogin = () => {
    let user = {email: this.state.email, password: this.state.password};
    this.fetchUserInfo(user);
    this.setState({email: '', password: '', error: false});
  }

  fetchUserInfo = user => {
    const options = {
      method: 'POST',
      body: JSON.stringify(user),
      headers: {
        'Content-Type': 'application/json'
      }
    }

    return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
      .then(response => response.json())
      .then(userInfo => this.props.addUser(userInfo.user))
      .catch(error => this.setState({error: true}))
  }

  render() {
    return (
      <form>
        <input
          type="text"
          placeholder="Email..."
          name="email"
          value={this.state.email}
          onChange={this.handleChange}
        />
        <input
          type="text"
          placeholder="Password..."
          name="password"
          value={this.state.password}
          onChange={this.handleChange}
        />
        <button className='login-btn' onClick={this.verifyInputs}>Login</button>
        {this.state.error ? <h2 className='error-msg'>Please enter a valid email and password!</h2> : null}
      </form>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  addUser: user => dispatch(updateUser(user))
})

export default connect(null, mapDispatchToProps)(LoginForm)
