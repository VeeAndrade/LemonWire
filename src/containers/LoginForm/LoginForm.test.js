import React from 'react';
import { LoginForm, mapDispatchToProps } from './LoginForm';
import { shallow } from 'enzyme';
import { updateUser, getRatings } from '../../actions';
import { getUserInfo, getUserRatings } from '../../apiCalls';

jest.mock('../../apiCalls')

describe('LoginForm', () => {
  describe('LoginForm container/component', () => {
    let wrapper;
    let mockEvent;
    let mockState;

    beforeEach(() => {
      wrapper = shallow(<LoginForm />)
      mockState = {
        email: 'fakeEmail',
        password: 'fakePassword',
        error: true
      }
      mockEvent = {
        target:
        {
          name: 'email',
          value: 'fake value'
        },
        preventDefault: jest.fn()
      }
      getUserInfo.mockImplementation(() => {
      return Promise.resolve([{ id: 29, name: 'Robbie' }]);
    })

      getUserRatings.mockImplementation(() => {
      return Promise.resolve([{ id: 300, movie_id: 20, rating: 5}]);
    });

  });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('should update state when handleChange is invoked', () => {
      expect(wrapper.state('email')).toEqual('')
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state('email')).toEqual('fake value')
    })

    it('should verify user inputs', () => {
      expect(wrapper.state('error')).toEqual(false)
      wrapper.instance().verifyInputs(mockEvent);
      expect(wrapper.state('error')).toEqual(true)
    });

    it('should call fetchUserInfo when completeLogin is invoked', () => {
      const mockState = {
        email: 'sam@turing.io',
        password: '123456',
        error: false
      }
      const expected = {
        email: 'sam@turing.io',
        password: '123456'
      }
      wrapper.instance().setState(mockState);
      wrapper.instance().fetchUserInfo = jest.fn()
      wrapper.instance().completeLogin()
      expect(wrapper.instance().fetchUserInfo).toHaveBeenCalledWith(expected)
    });

    it('should clear state when complete login is invoked', () => {
      const defaultState = {
        email: 'sam@turing.io',
        password: '123456',
        error: false
      }
      const expected = { email: '', password: '', error: false };
      wrapper.instance().setState(defaultState);
      wrapper.instance().completeLogin();
      expect(wrapper.state()).toEqual(expected);
    });

    it('should fire off getUserInfo when fetchUserInfo is invoked', () => {
      const mockUser = {name: 'Robbie'}
      wrapper.instance().fetchUserInfo(mockUser)
      expect(getUserInfo).toHaveBeenCalledWith(mockUser)
    });

    it('should fire off getUserRatings when fetchUserRatings is invoked', () => {
      const mockUserId = 7
      wrapper.instance().fetchUserRatings(mockUserId)
      expect(getUserRatings).toHaveBeenCalledWith(mockUserId)
    });

    it('should fire off verifyInputs when login button is clicked', () => {
    wrapper.instance().verifyInputs = jest.fn();
    wrapper.instance().forceUpdate()
    const mockEvent = {
      preventDefault: jest.fn()
    }
    wrapper.find('.login-btn').simulate('click', mockEvent)
    expect(wrapper.instance().verifyInputs).toHaveBeenCalled()
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the updateUser action when fetchUserInfo is called', () => {
      let mockResponse = { user : {
          id: 1,
          name: 'Robbie'
      }};
      window.fetch = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockResponse)
        });
      });
      const mockDispatch = jest.fn();
      const actionToDispatch = updateUser(mockResponse);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addUser(mockResponse)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

  it('should call dispatch with the getRatings action when getUserRatings is called', () => {
    let mockResponse = { ratings : {
        movie_id: 10,
        rating: 6
    }};
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockResponse)
      });
    });
    const mockDispatch = jest.fn();
    const actionToDispatch = getRatings(mockResponse);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addUserRatings(mockResponse)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })
});
