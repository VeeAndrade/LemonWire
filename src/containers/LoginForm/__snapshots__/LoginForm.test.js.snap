// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm LoginForm container/component should match snapshot 1`] = `
<form>
  <input
    name="email"
    onChange={[Function]}
    placeholder="Email..."
    type="text"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="Password..."
    type="text"
    value=""
  />
  <button
    className="login-btn"
    onClick={[Function]}
  >
    Login
  </button>
</form>
`;
